Document JSON::PullParser

erdnaxeli opened the pull request:
Hi, 

I wanted to contribute to crystal, and I found the issue #8050 wich is labeled as "community:newcomer". English is not my native language so there are probably mistakes or bad sentences.

I ran the documentation examples, they should all be valid.

erdnaxeli commented:
It shoul fix #9983 too.

straight-shoota commented:
This is really great! I'm not a native speaker myself, but the wording seems pretty well 👍 There are however a number of small typos. Would you mind maybe running a spell-checker in your code editor? That should fix this easier than pointing each one out.

erdnaxeli commented:
It should be better now.

PS : What do you prefer when editing commits? I amended it and used push force, but I can push new commits and squash them when the PR is ok.

straight-shoota commented:
Please just add new commits and don't do force pushes. This makes reviewing much easier.
The commits will be squashed on merge anyway.

straight-shoota commented on src/json/pull_parser.cr:
```diff
@@ -1,4 +1,52 @@
 # This class allows you to consume JSON on demand, token by token.
+#
+# Each *read_** method consumes the next token.
```
In most markdown parsers this should be `*read_\**` but I'm not sure how the docs generator renders it. Did you preview this?

erdnaxeli replied:
I tried with `*read_\**`, but it is rendered as _read\_\\*_. Surprisingly `*read_**` just works.

straight-shoota replied:
Okay, it's fine then.

straight-shoota commented on src/json/pull_parser.cr:
```diff
@@ -87,20 +145,28 @@ class JSON::PullParser
 
+  # Reads a whole object.
+  #
+  # It reads the beginning of the object, yield each key and key location, and reads the end of the object.
```
Maybe add that the values need to be consumed in the block, otherwise the pull parser breaks?

straight-shoota commented on src/json/pull_parser.cr:
````diff
@@ -138,6 +210,21 @@ class JSON::PullParser
+  # If the value in unknown, it raises a `ParseException`.
+  #
+  # ```
+  # pull = JSON::PullParser.new(input)
````
I'd suggest parsing an array of differently-typed values as a self-contained example of all read_raw results:
```cr
pull = JSON::PullParser.new %([nil, true, 1, "foo", [1, "two"], {"foo": "bar"}])
pull.read_begin_array
pull.read_raw # => nil
pull.read_raw # => true
pull.read_raw # => 1
pull.read_raw # => "foo"
pull.read_raw # => "[1, \"two\"]"
pull.read_raw # => "{\"foo\": \"bar\"}"
pull.read_end_array
```

straight-shoota replied:
The values are actually all of type `String`.

straight-shoota commented on src/json/pull_parser.cr:
```diff
@@ -233,12 +331,18 @@ class JSON::PullParser
 
+  # Reads an object's keys and yield when *key* is found.
+  #
+  # All the other object's keys are skipped.
```
```suggestion
  # All the other object keys are skipped.
```

straight-shoota commented on src/json/pull_parser.cr:
```diff
@@ -259,42 +363,62 @@ class JSON::PullParser
     @kind
   end
 
+  # Reads a boolean value and returns it. If the value is not a boolean, returns nil.
```
```suggestion
  # Reads a `Bool` value and returns it. If the value is not a boolean, returns nil.
```

Could you add backticks to the types below as well? That creates links to their API docs.

straight-shoota commented on src/json/pull_parser.cr:
```diff
@@ -399,20 +523,29 @@ class JSON::PullParser
+  # Returns the current location.
+  #
+  # The location is a tuple {line number, column number}.
   def location
```
```suggestion
  def location : Tuple(Int32, Int32)
```

asterite commented:
I wonder if we should make the pull parser be more type safe. Like having the value be a union of all possible types...

straight-shoota commented:
@asterite How do you mean? All read_* methods return the respective data type: for example `read_int` returns `Int64`.

erdnaxeli commented:
I fixed what you commented about @straight-shoota. I also added an example of a failure in top documentation.

I am not sure to understand what you mean @asterite .

asterite commented:
What I mean is, there's a `kind` method (getter) that returns what kind of token it is. If it's a bool, then you get the value by calling `bool_value`. If it's an int you get the value by calling `int_value`. The thing is, if it's a bool and you call `int_value`, you get an int, but it's garbage (or it can't be relied). So you can misuse the pull parser.

I did it like that because I think it's just more efficient, but I don't know.

A more type-safe way would be to replace `kind` with `value` of type `JSON::PullParser::Value`, like this:

```crystal
alias Value = Null | BoolValue | IntValue | FloatValue | StringValue | BeingArray | EndArray | BeingObject | EndObject | EOF

record Null
record BoolValue, value : Bool
record IntValue, value : Int64
# and so on...
```

That way when you can never read an `int` out of a `BoolValue`.

That said, it seems `kind` and all those `*_value` are more of low-level primitives for the pull parser and the usual way to interact with it is through the `read_` methods which are fine, so we could make it better after 1.0, though it's a breaking change.

asterite commented:
Also, this PR is fine, no need to do anything... I just wanted to introduce the discussion but it probably belongs somewhere else...

straight-shoota commented:
Thank you @erdnaxeli
